### ISSUE TITLE
Target Allocator - ServiceMonitor scheme

### ISSUE THREAD

**Issue by rhysxevans** [open-telemetry/opentelemetry-operator#1669]
Hi

I have the TA up and with the `PrometheusCR` enabled, this is to look at options of migrating to using the OTEL collector for more purposes.

At present I am getting data as expected except for service monitors (to be fair in my test lab I only have a couple of these setup) that contain a "non default" scheme and have authentication data in them. eg

```
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  creationTimestamp: "2023-04-06T18:07:21Z"
  generation: 1
  labels:
    app: first-cluster
  name: first-cluster
  namespace: opensearch-first-cluster
  resourceVersion: "991661"
  uid: cfad90db-7361-4b05-96de-b89b36537dbc
spec:
  endpoints:
  - basicAuth:
      password:
        key: password
        name: first-cluster-opensearch-monitoring
      username:
        key: username
        name: first-cluster-opensearch-monitoring
    interval: 30s
    path: /_prometheus/metrics
    port: http
    scheme: https
    tlsConfig:
      insecureSkipVerify: true
  jobLabel: opensearch_first-cluster
  namespaceSelector:
    matchNames:
    - opensearch-first-cluster
  selector:
    matchLabels:
      opster.io/opensearch-cluster: first-cluster
```

```
apiVersion: v1
kind: Service
metadata:
  annotations:
    banzaicloud.com/last-applied: UEsDBBQACAAIAAAAAAAAAAAAAAAAAAAAAAAIAAAAb3JpZ2luYWyUks2O2zAMhN+FZ3nrrGO362O39wRN0UuRAy3TjRBFEig6PQR690JO4Py0QLA3ezTzYUjwBBjMT+JovIMWjgtQsDeuhxY2xEejCRQcSLBHQWhPYLEjG/OXD1GIX4z/5AO5SMh6V2g7ZhVaGAxHmf+TAocH+kc/yzGgzm83oEeb/+OIv9NATE5ThPbXQ/Vr9uXabJqns17vVzn/jSzJZBceSYH2Tthbmwuflcvsq0BuM8He74r+p/9ocqBrqhL1siwWr8NbsWw+98WXCrFoqnro+mW5oK6GtE0KYiCd1xc8y3mMC3gnEkBNOrRvr2WpILAXr72FFn68r0GBIP8mWc+WpOa4MLo4hWdG9ZxR3TEOJGx0vBKa54TmjsAab9L183Rdpq2CSJa0eP74XeWNCso4naT12H9Fi05n6yml9DcAAP//UEsHCDhHq9BCAQAA4AIAAFBLAQIUABQACAAIAAAAAAA4R6vQQgEAAOACAAAIAAAAAAAAAAAAAAAAAAAAAABvcmlnaW5hbFBLBQYAAAAAAQABADYAAAB4AQAAAAA=
  creationTimestamp: "2023-04-06T18:07:23Z"
  labels:
    opster.io/opensearch-cluster: first-cluster
  name: first-cluster
  namespace: opensearch-first-cluster
  ownerReferences:
  - apiVersion: opensearch.opster.io/v1
    blockOwnerDeletion: true
    controller: true
    kind: OpenSearchCluster
    name: first-cluster
    uid: b630ac40-12f9-467d-83aa-635fbd401eb5
  resourceVersion: "991678"
  uid: add9fb87-651a-4898-8a10-ccc29ed0b043
spec:
  clusterIP: 192.168.5.15
  clusterIPs:
  - 192.168.5.15
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - name: http
    port: 9200
    protocol: TCP
    targetPort: 9200
  - name: transport
    port: 9300
    protocol: TCP
    targetPort: 9300
  - name: metrics
    port: 9600
    protocol: TCP
    targetPort: 9600
  - name: rca
    port: 9650
    protocol: TCP
    targetPort: 9650
  selector:
    opster.io/opensearch-cluster: first-cluster
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}
```

This config works with prometheus but using the otel collector (via the TA) I get `2023-04-20T20:04:56.248Z	warn	internal/transaction.go:121	Failed to scrape Prometheus endpoint	{"kind": "receiver", "name": "prometheus", "data_type": "metrics", "scrape_timestamp": 1682021096246, "target_labels": "{__name__=\"up\", container=\"opensearch\", endpoint=\"http\", instance=\"100.64.141.115:9200\", job=\"first-cluster-masters\", namespace=\"opensearch-first-cluster\", pod=\"first-cluster-masters-1\", service=\"first-cluster-masters\"}"}`

Just looking to find out if this is a known "issue" or is, and it probably is, it that I am doing something wrong? Any help is appreciated

This is all being deployed using the opentelemetry-operator helm chart version 0.26.3 

Target Allocator
```
apiVersion: v1
data:
  targetallocator.yaml: |
    allocation_strategy: least-weighted
    config:
      scrape_configs:
      - job_name: otel-collector
        scrape_interval: 10s
        static_configs:
        - targets:
          - 0.0.0.0:8888
    label_selector:
      app.kubernetes.io/component: opentelemetry-collector
      app.kubernetes.io/instance: otel.allocator
      app.kubernetes.io/managed-by: opentelemetry-operator
kind: ConfigMap
metadata:
  creationTimestamp: "2023-04-20T18:37:48Z"
  labels:
    app.kubernetes.io/component: opentelemetry-targetallocator
    app.kubernetes.io/instance: otel.allocator
    app.kubernetes.io/managed-by: opentelemetry-operator
    app.kubernetes.io/name: allocator-targetallocator
    app.kubernetes.io/part-of: opentelemetry
    app.kubernetes.io/version: latest
  name: allocator-targetallocator
  namespace: otel
  ownerReferences:
  - apiVersion: opentelemetry.io/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: OpenTelemetryCollector
    name: allocator
    uid: 012622e4-db73-4c9c-8f7c-39e47cdabb8f
  resourceVersion: "9749302"
  uid: a591df05-6f7e-4267-8953-338363472f87
```

Collector
```
apiVersion: v1
data:
  collector.yaml: |
    exporters:
      logging:
        loglevel: debug
      prometheusremotewrite:
        endpoint: http://prometheus.kube-prometheus-stack.svc.cluster.local:9090/api/v1/write
        external_labels:
          scraper: ${POD_NAME}
          source: otel-allocator
    extensions:
      memory_ballast:
        size_in_percentage: 20
    processors:
      batch:
        send_batch_max_size: 1000
        send_batch_size: 800
        timeout: 30s
      memory_limiter:
        check_interval: 1s
        limit_percentage: 70
        spike_limit_percentage: 30
    receivers:
      hostmetrics:
        collection_interval: 30s
        scrapers:
          cpu: null
          load:
            cpu_average: true
          memory: null
      prometheus:
        config:
          global:
            scrape_interval: 1m
            scrape_timeout: 10s
            evaluation_interval: 1m
          scrape_configs:
          - job_name: otel-collector
            honor_timestamps: true
            scrape_interval: 10s
            scrape_timeout: 10s
            metrics_path: /metrics
            scheme: http
            follow_redirects: true
            enable_http2: true
            http_sd_configs:
            - follow_redirects: false
              enable_http2: false
              url: http://allocator-targetallocator:80/jobs/otel-collector/targets?collector_id=$POD_NAME
        target_allocator:
          collector_id: ${POD_NAME}
          endpoint: http://allocator-targetallocator
          http_sd_config:
            refresh_interval: 60s
          interval: 30s
    service:
      extensions:
      - memory_ballast
      pipelines:
        metrics:
          exporters:
          - prometheusremotewrite
          processors:
          - memory_limiter
          - batch
          receivers:
          - prometheus
kind: ConfigMap
metadata:
  creationTimestamp: "2023-04-20T18:37:48Z"
  labels:
    app.kubernetes.io/component: opentelemetry-collector
    app.kubernetes.io/instance: otel.allocator
    app.kubernetes.io/managed-by: opentelemetry-operator
    app.kubernetes.io/name: allocator-collector
    app.kubernetes.io/part-of: opentelemetry
    app.kubernetes.io/version: latest
  name: allocator-collector
  namespace: otel
  ownerReferences:
  - apiVersion: opentelemetry.io/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: OpenTelemetryCollector
    name: allocator
    uid: 012622e4-db73-4c9c-8f7c-39e47cdabb8f
  resourceVersion: "9771170"
  uid: db4349e1-0f44-4b71-af3e-7467b80b7696
```

**Comment by rhysxevans** [comment]
So I did some digging and I think this is not the `scheme` more the access to the secret that hosts the login details for the service to be monitored

**Comment by jaronoff97** [comment]
Could you send the response from the target allocator's scrape_configs endpoint? (i.e. `curl http://allocator-targetallocator:80/scrape_configs`) This would help us ensure that the configuration for this is being propagated correctly. I imagine/wonder if the TA needs to learn how to pull secret data from ServiceMonitor configuration if the prometheus-operator doesn't do that automatically. Could you also send any target allocator logs that exist?

**Comment by rhysxevans** [comment]
Service Monitor setup

```
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  creationTimestamp: "2023-04-06T18:07:21Z"
  generation: 1
  labels:
    app: first-cluster
  name: first-cluster
  namespace: opensearch-first-cluster
  resourceVersion: "991661"
  uid: cfad90db-7361-4b05-96de-b89b36537dbc
spec:
  endpoints:
  - basicAuth:
      password:
        key: password
        name: first-cluster-opensearch-monitoring
      username:
        key: username
        name: first-cluster-opensearch-monitoring
    interval: 30s
    path: /_prometheus/metrics
    port: http
    scheme: https
    tlsConfig:
      insecureSkipVerify: true
  jobLabel: opensearch_first-cluster
  namespaceSelector:
    matchNames:
    - opensearch-first-cluster
  selector:
    matchLabels:
      opster.io/opensearch-cluster: first-cluster
```

**Comment by rhysxevans** [comment]
`first-cluster-opensearch-monitoring` is a secret

**Comment by rhysxevans** [comment]
It is not a valid one

**Comment by rhysxevans** [comment]
Ok, let me know if you need me to do anything. And thanks for your help

**Comment by jaronoff97** [comment]
Alright, i found the problematic lines! We aren't setting the secrets in [the store](https://github.com/open-telemetry/opentelemetry-operator/blob/main/cmd/otel-allocator/watcher/promOperator.go#L158-L164) which is what gets used by the prometheus-operator generation [here](https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/prometheus/promcfg.go#L1133) and [here](https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/prometheus/promcfg.go#L323-L325). I believe the fix is going to be in setting the store, though i'm not sure how to go about that right now.

**Comment by jaronoff97** [comment]
ah okay, prometheus-operator has a bunch of functions we need to call to make this work something like [this block](https://github.com/prometheus-operator/prometheus-operator/blob/e753149c76e60ca45156d8293625f3f732869f5f/pkg/prometheus/server/operator.go#L1561-L1577). I'd say this work is possible, but not a tiny change... I don't have a ton of capacity to work on it currently unfortunately. I'll ask around in the next operator SIG meeting and in the slack and see if anyone else can take it.

**Comment by rhysxevans** [comment]
Ok, thanks for you help

**Comment by jaronoff97** [comment]
@matej-g offered to help out here, thank you! Please let me know if you need any more clarification here.

**Comment by matej-g** [comment]
Some additional context, I also chatted with @jaronoff97.

It seems that currently, we won't be able to support specifying these credential fields that are coming from secrets, without further work, possibly on both target allocator and receiver.

When we specify a secret in the scrape config, it will get redacted to `<secret>` in the `/scrape_configs` response (see my screenshot here - https://github.com/open-telemetry/opentelemetry-operator/pull/1710#issuecomment-1534870046), making it not possible for receiver to know the actual secret. I think we'd first need to find a way to communicate which secret needs the collector look into in order to properly build credentials on the collector side of things.

On the other hand, this should not affect credentials that can be provide with a file path - this would only require that the collector has that particular secret mounted.

**Comment by matej-g** [comment]
Sorry for the silence on this one. It looks like we're not the only project dealing with this issue, potentially we might have a solution from https://github.com/prometheus/common/pull/487 (we might have to warn our users that their secrets might be exposed via target allocator endpoint, alas that is the downside if they want to use credentials).

In order to move forward with #1710 (it also includes addition of some unit tests which are useful outside of this issue as well), @jaronoff97 would you be fine going with your suggestion - i.e. we can state that for now we support only those types of credentials that can be provided in a file, document this and revisit this topic once https://github.com/prometheus/common/pull/487 moves forward.

This will unfortunately not unblock the present issue instantly, but allows us to at least provide partial solution.

**Comment by matej-g** [comment]
I think this was mistakenly marked as resolved, #1710 provided only partial solution, this should be re-open.

**Comment by matej-g** [comment]
So this popped up again on my radar and I would like to address this, together with #1844 and build on changes incoming in #2328. 

But I think we need some alignment on the next steps. It seems we will need to deal with two distinct use cases:
- _Mounting files for authorization / TLS_ - These should be mounted by the collector, so when they are referenced in the endpoint config, collector pod(s) can use them. We could require users to mount it themselves, as suggested previously, or to be more user friendly, the operator should ensure that these are mounted automatically.

- _Referencing secret / configmap_ - As I described above, the problem here is that we need to communicate the content of the secret between the TA and collector. But default, based on the upstream Prometheus code,  a secret is being redacted during marshalling, so when collector pulls the scrape configs, it will only show up as `<secret>`. I see two ways to work around this:
  1. Override the marshalling and expose the secret values (a la solution referenced here https://github.com/prometheus/common/pull/487#issue-1709617209), making the collector capable of directly "reading" it. Obviously, this opens up the possibility of leaking a secret, since they will be present in plain text in the `scrape_config`.
  2. Instead of providing secrets directly, communicate which secrets / configmaps keys and values should be used for particular endpoint authorization. The collector would then need to pull the secret and get the actual value, and use that in the authorization config. This would require more work and adjustments on the collector, but would avoid having to provide secrets directly.
  
 cc @jaronoff97

**Comment by jaronoff97** [comment]
@matej-g i've been thinking about this as well... let me address each of the cases:
1. mounting is fine, i think the issue is that the operator may not know about the configs that need to be mounted. The scrape configs generated by the servicemonitors come from the TA directly to the collector, which means the operator isn't aware of what needs to be present. 
2. I think the first option is going to be easier architecturally than the second. The second option has the same problem as the first which is that the operator isn't aware of what secrets / configmaps need to be mounted.

All of this brings me to an unfortunate conclusion... I think we are reaching the limit of what is possible with this servicemonitor CRD architecture. The decision to have the CRDs be pulled by the target allocator was one that was made prior to me joining the project and it didn't feel right to critique it at the time. Given the issue we are discussing is one of many that stems from this architecture choice and that we are discussing the next version of our own architecture, I think now is the time to figure out if there is a better way to architect this functionality.

In my opinion, we should move the CRD functionality entirely to the operator. The operator would take on the onus to pull the prometheus CRDs, translate their YAML to a scrape config and simply write it as part of the collector's (and TA's) config map. This would allow the operator to provide much more convenience when it comes to secrets and configs because we would no longer need to deal with the secrets being marshalled. Furthermore, moving this functionality out of the target allocator and on to the operator would reduce the scope of the TA (and probably improve its performance) and improve the overall user experience for the prom CRDs. The only drawback of doing this is that non-operator users would no longer be able to take advantage of this. Given that doing so has always been done outside our recommendations, I think this is a worthwhile tradeoff. ex:

```mermaid
sequenceDiagram
    User->>+Operator: Applies collector w/ promCRD enabled
    Operator->>+Operator: Reconciles collector CRD
    Operator-->>-Operator: Pulls prom CRDs that match selector
    Operator->>+Target-Allocator: Creates TA w/ CRD config
    Operator->>+Collector: Creates Collector w/ CRD config
    Collector->>Target-Allocator: Requests Targets
```

On any change from the operators prom crd watcher it simply rolls out a new version of the collector and target allocator which both will have a hash of the scrape config as an annotation. Slotting in to the above configuration for the issue at hand would look like this:

```mermaid
sequenceDiagram
    User->>+Operator: Applies collector w/ promCRD enabled
    Operator->>+Operator: Reconciles collector CRD
    Operator-->>-Operator: Pulls prom CRDs that match selector
    Operator->>+User: Propagates warnings for missing secrets/config mounts
```

After this exists, we could visit the possibility of doing this automatically for a user as well, though I'm sure that's going to be a bit thornier. Let me know your thoughts! Thank you :bow: 

cc @swiatekm-sumo who may have some better ideas about this.

**Comment by swiatekm-sumo** [comment]
For reference, what prometheus-operator does here is simply writing the whole configuration, credentials included, into a Secret, which is then mounted in the Prometheus Pod and config-reloader takes care of making Prometheus load the new configuration. I mention this because if we want to do anything outside of this loop, we'll need to add it ourselves.

A similar architecture, which I understand Jacob's proposal to be, requires that we deal with the configuration reload problem, which is quite thorny, and is the reason prometheus-operator uses a config-reloader program in the first place. Restarting everything every time this configuration changes is a brute force solution, and I don't think it would scale well to large clusters. It certainly feels wrong to me to force these restarts on users who don't need them (because they don't need autorization for their scrapes). The alternative is to use something like config-reloader, but we'd need it for both the collector and the target allocator, and from what I know it's relatively expensive to reload the collector.

Making users mount the Secrets manually in the Collector seems like it could be a reasonable workaround for the time being, but we'd also need to do develop a convention for file naming, as ServiceMonitor only allows a Secret name to be specified for authorization, so Target Allocator would need to change these to file path references, and the user would actually need to mount the Secrets under these file paths.

Maybe the mid-term solution is simply to secure the connection between the Collector and Target Allocator? It wouldn't be particularly difficult to set up client TLS authentication for this connection - the operator can easily generate the Secret and ensure it's mounted in all the Pods. This also has some negative consequences, like making troubleshooting TA harder, but along with custom unmarshaling, it solves all the problems, and is only marginally more complicated than the filename approach.

**Comment by jaronoff97** [comment]
Yes, the configuration reload problem is definitely thorny. I was thinking with the rate limiting you put in place for the TA, we would use that to limit how often we're rolling out new workloads. 

I think setting up TLS between the Collector and TA is a good thing to do regardless of this issue and should be possible given we have a dependency on cert-manager already. 

Let's discuss these ideas at next week's SIG meeting.

**Comment by swiatekm-sumo** [comment]
> Yes, the configuration reload problem is definitely thorny. I was thinking with the rate limiting you put in place for the TA, we would use that to limit how often we're rolling out new workloads.

I don't think any amount of rate limiting will fix this. We're adding a per-node allocation strategy in #2430, and when using that you'd have to recreate all the Pods of a DaemonSet, which can be hundreds in a large cluster. Putting the usability of this aside, users have come to expect performance similar to prometheus-operator, which takes under a minute from a ServiceMonitor change to the first new scrape.

**Comment by jaronoff97** [comment]
yeah that's very true. Ugh. I can check in with collector SIG people and see if their efforts for dynamic reloading have made progress.

**Comment by matej-g** [comment]
I won't be able to join the SIG meeting unfortunately, but I'm interesting to hear the outcomes.

**Comment by ItielOlenick** [comment]
Coming from https://github.com/open-telemetry/opentelemetry-operator/blob/main/cmd/otel-allocator/README.md#service--pod-monitor-endpoint-credentials which mentions: "your collector instance needs to have the particular secret mounted as a file at the correct path".

How would one achieve this? What is the _correct path_ at the collector instance?

**Comment by swiatekm-sumo** [comment]
@ItielOlenick it is whatever you set it to be in your configuration. For example, if you set `bearerTokenFile` in a ServiceMonitor to `/var/secrets/bearer_token`, then you need to ensure that file exists in the Collector Pods.

**Comment by ItielOlenick** [comment]
Isn't that specific for bearer auth? What about basic auth?

**Comment by swiatekm-sumo** [comment]
You can't currently do it for basic auth, for reasons discussed in this thread. ServiceMonitors only let you define that as a reference to a Secret, and the serialized value of this secret won't be sent to the collector.

The mid-term plan to fix this is to secure the collector <-> target allocator connection using client-side TLS and use custom serialization for Prometheus config to send the credential values over HTTPS. I don't think anyone is actively working on it right now, though.

**Comment by alita1991** [comment]
I hit this issue today while trying to find a solution to scrape protected metrics endpoints, I created a ServiceMonitor with basic_auth credentials and the collector is unable to scrape the endpoint.

Still waiting for https://github.com/prometheus/common/pull/487 to be merged, I know is not the perfect solution, but can unblock me. 

My goal is to reduce the manual work on the collectors and the ServiceMonitor/PodMonitor scrape is the key solution.

**Comment by swiatekm-sumo** [comment]
FYI @alita1991 I'm fine using the kind of "horrible workaround" mentioned in that PR. But in order to actually do this, we'd need to secure the connection between the collector and operator. My idea for implementing this was:

* Add an optional https server to the target allocator
* Generate a TLS cert for the target allocator
* Mount the TLS cert to the collector, use it for authenticating the connection to the target allocator
* Mount the TLS cert to the target allocator, set up client-side TLS auth
* If serving over https, use custom marshalling to include secrets in the scrape config

**Comment by rashmichandrashekar** [comment]
> FYI @alita1991 I'm fine using the kind of "horrible workaround" mentioned in that PR. But in order to actually do this, we'd need to secure the connection between the collector and operator. My idea for implementing this was:
> 
> * Add an optional https server to the target allocator
> * Generate a TLS cert for the target allocator
> * Mount the TLS cert to the collector, use it for authenticating the connection to the target allocator
> * Mount the TLS cert to the target allocator, set up client-side TLS auth
> * If serving over https, use custom marshalling to include secrets in the scrape config

@swiatekm-sumo -Will this be available anytime soon? Since the latest version of prometheus common with the fix is picked up by the operator?

**Comment by swiatekm-sumo** [comment]
@rashmichandrashekar I don't think anyone is actively working on this at the moment. If you or anyone else would like to, I can provide more detailed guidance on how to proceed.

**Comment by ItielOlenick** [comment]
@swiatekm-sumo I'd like to work on this feature

**Comment by swiatekm-sumo** [comment]
Great, thanks for picking this up! My suggestion would be to start by adding the https server, then enable communication over mtls between the collector and target allocator, and finally the secret marshalling change. All of this together sounds like a lot for a single PR, and smaller PRs get reviewed faster on average.

**Comment by ItielOlenick** [comment]
@swiatekm-sumo After familiarizing myself with the codebase, I've started adding the HTTPS server and experimented with the functionality introduced in https://github.com/prometheus/common/pull/487.

I have some questions and ideas:

1. It makes sense to have only one of the HTTPS/HTTP servers running, since we probably don't want to keep two versions of the scrape_config (one with the secrets revealed and one without) plus the rest of the server's logic.
2. Readiness and liveness probes - since we are serving with mTLS, we cannot use the `HTTPGetAction` even with the schema as HTTPSs. I'm thinking of either creating a small executable that will make use of the mounted certs and make the necessary call to the server using `ExecAction`, or creating a separate server dedicated for the probes.

**Comment by swiatekm-sumo** [comment]
I actually wanted to keep both servers, because hiding everything behind auth makes debugging a lot more difficult. One of the simplest ways of troubleshooting issues with the target allocator is to forward the http port and just check the endpoints in a web browser. Probes are another reason to keep the http server.

Is it a big problem to keep both versions of the serialized scrape configs?

**Comment by ItielOlenick** [comment]
I agree.
It will be simpler to have both servers running from that perspective. I'll look into that.

**Comment by ItielOlenick** [comment]
Regarding the next steps, this is what i had in mind:
* Add logic to the collector to enable using mTLS when connecting to the TA
* Take advantage of the existing cert manager to manage the CA, server and client certs. Should we reuse the existing issuer already configured?
* Add relevant flags/configs to mount certs to both client and server and use mTLS

**Comment by swiatekm-sumo** [comment]
> Regarding the next steps, this is what I had in mind:
> 
>     * Adding logic to the collector to enable using mTLS when connecting to the TA.

This was already done in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31449, we just need to configure the receiver correctly.

>     * Taking advantage of the existing certificate manager to manage the CA, server, and client certificates. Should we reuse the existing issuer already configured?

This might be a can of worms. Technically, we require cert-manager, but this is only for webhooks. The Operator Helm chart has an option to generate the certs statically without cert-manager, so if we put a hard dependency on it here, we'll be locking those users out of this functionality. I'd prefer not to do that, if possible, but that decision depends on how complex the non-cert-manager implementation will be. @jaronoff97 @pavolloffay wdyt?

>     * Adding relevant flags/configurations to mount certificates to both the client and server and utilize mTLS.

:+1:

**Comment by jaronoff97** [comment]
I definitely agree with the above for the first and third points... for the middle point, I think we could take a similar to approach as we now do for the autodetection for RBAC permissions. i.e. if you give the operator permissions for the creating and managing certs AND you have cert manager installed, we will automatically secure the connection for you. I don't think this needs to be in the initial version of this and until then we could simply provide some example configs on how this would be done.

**Comment by swiatekm-sumo** [comment]
That's fair enough for an initial implementation, I agree.

**Comment by thefirstofthe300** [comment]
It looks like https://github.com/open-telemetry/opentelemetry-operator/pull/2921 adds support for scraping real values from the target allocator. Is the operator going to need cert manager to generate the required certificate? Is there going to be some documentation going out on how to use this feature?

**Comment by ItielOlenick** [comment]
For the initial version i was thinking of this addition to the CRD API:
```yaml
spec:
  config:
    targetAllocator:
      mtls:
        enabled: true
        targetAllocatorCerts:
          ca_file: ""
          cert_file: ""
          key_file: ""
        collectorCerts:
          ca_file: ""
          cert_file: ""
          key_file: ""
```
What do you think?

**Comment by thefirstofthe300** [comment]
My two cents would be that the underscores should be removed and replaced with lowerCamelCasing. 

From the end-user perspective, what purpose do these files serve? Are they references to files mounted by a secret? If so, how would I define the secret? If they're managed by the OTEL operator, why do I need to care about the certificate configuration at all?

**Comment by ItielOlenick** [comment]
As mention above, since the secrets for the initial iteration will not be
managed by the operator, the users will be creating the secrets and
mounting them using the specs for the TA and the Collector.

**Comment by swiatekm-sumo** [comment]
I was actually thinking that we should start with what Jacob mentioned - if the cert-manager CRDs are installed and the operator has permission to create them, then use cert-manager to provision the certificates. I'd rather avoid adding fields to the Collector CRDs until we're confident we know how this should work. And ideally, I'd rather avoid making this configurable at all.

**Comment by ItielOlenick** [comment]
Since @jaronoff97 suggested the initial version of this change doesn't need to include managing certs and we will provide example configs in the meanwhile, i was going in this direction.
Open to suggestions about how we should get the relevant configs to the Collector/TA without changing the CRDs. Unless we do want to go with managing certs with cert-manager off the bat.

**Comment by swiatekm-sumo** [comment]
I'd personally rather do the no-config version first, which means using cert-manager. There isn't really any benefit to the user orchestrating the certs themselves, other than us not needing to do it, and like I mentioned, I really don't want to add fields to the CRD unless absolutely necessary.

**Comment by ItielOlenick** [comment]
If the [readme](https://github.com/open-telemetry/opentelemetry-operator?tab=readme-ov-file#getting-started) specifically states that cert manager needs to be installed in the cluster, why do we need to check if the CRDs exist for this functionality?

**Comment by swiatekm-sumo** [comment]
That, in my opinion, is a documentation failure we should fix. Cert-manager is used by the default kustomize manifests for provisioning webhook certificates, but it's very much possible to not use it. For example, the official Helm Chart can work without cert-manager by manually creating the necessary Secrets.

I don't think we should be making the dependency on cert-manager deeper.

**Comment by ItielOlenick** [comment]
Got it. Thanks!

**Comment by ItielOlenick** [comment]
Done the initial implementation. Tested locally with curl, open_ssl, etc.

Testing with the Collector i see that although https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31449 was merged the Collector client doesn't provide the certificates with the requests to the TA.

Opened a bug - https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33370

**Comment by ItielOlenick** [comment]
@swiatekm-sumo whenever there is a basicAuth in the spec.endpoints for a serviceMonitor the /jobs is empty although /scrape_configs returns the expected result.

No error in the logs of the TA.
Any ideas as to where to look for the issue?
Other serviceMonitors with out basicAuth work as expected.

**Comment by swiatekm-sumo** [comment]
That's weird. Targets are emitted by Prometheus' own discovery manager, so maybe we're not using it correctly? If scrape_configs is correct, then the config generation at least works correctly.

**Comment by swiatekm-sumo** [comment]
Are you sure the Target Allocator has RBAC permissions to read the required Secrets? I'm not sure if we'd get an error about that.

**Comment by ItielOlenick** [comment]
In the /scrape_configs endpoint I see the secret (when working with https). Double checked the rbac rules. There is get on all secrets

**Comment by ItielOlenick** [comment]
You are not using TLS between the TA and the Collector. you should be seeing <secret> as before  https://github.com/open-telemetry/opentelemetry-operator/pull/2921 in the scrape_config.
Im seeing this issue when TA and Collector are communicating over mTLS.

**Comment by swiatekm-sumo** [comment]
So you're missing jobs, but only over https? They exist over http?

**Comment by ItielOlenick** [comment]
No. now even over http they are missing

**Comment by swiatekm-sumo** [comment]
Must be a problem with the http server then, as the targets are clearly generated. You're the foremost expert on this, I think, I can't see anything obviously wrong with the https server setup. Maybe we can catch this via unit tests?

**Comment by ItielOlenick** [comment]
Also, though you have basicAuth the actual /metrics endpoint doesn't require it.
Using your serviceMonitor work for me now.

In not exactly 100% on the flow, but could it be that if basicAuth is there and should be used at the endpoint but fails somehow, no job will be created?

**Comment by ItielOlenick** [comment]
I tried adding basicAuth to a serviceMonitor for which the TA does create a job (no auth is required at the /metrics endpoint) and it works as well. but as soon as i have a serviceMonitor for a service which requires basicAuth and the basicAuth is present in the ServiceMonitor, the job doesn't get created.
I will try some more variations but im running out of ideas :/

**Comment by swiatekm-sumo** [comment]
Can we move this conversation to the CNCF Slack? This back and forth is clogging the issue, and it sounds like we need to really dive into the weeds for this one.

**Comment by ItielOlenick** [comment]
@swiatekm-sumo Added https://github.com/open-telemetry/opentelemetry-operator/pull/3015 draft. 
Pretty big change, would appreciate a review before i move on with writing tests.

**Comment by jaronoff97** [comment]
@ItielOlenick is this work completed now? I think we got in everything we need now right?

**Comment by ItielOlenick** [comment]
> @ItielOlenick is this work completed now? I think we got in everything we need now right?

I believe so, yes.

**Comment by jaronoff97** [comment]
Awesome, thank you so much for your incredible work here. For any on-lookers, please open _new_ issues with relation to Target Allocator <> mTLS.
